src/middleware/identity.rs

```diff
@@ -284,84 +286,133 @@ where
 
 struct CookieIdentityInner {
     key: Key,
+    key_v2: Key,
     name: String,
     path: String,
     domain: Option<String>,
     secure: bool,
     max_age: Option<Duration>,
     same_site: Option<SameSite>,
+    visit_deadline: Option<Duration>,
+    login_deadline: Option<Duration>,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+struct CookieValue {
+    identity: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    login_timestamp: Option<SystemTime>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    visit_timestamp: Option<SystemTime>,
+}
+
+#[derive(Debug)]
+struct CookieIdentityExtention {
+    login_timestamp: Option<SystemTime>
 }
 
 impl CookieIdentityInner {
     fn new(key: &[u8]) -> CookieIdentityInner {
+        let key_v2: Vec<u8> = key.iter().chain([1, 0, 0, 0].iter()).map(|e| *e).collect();
         CookieIdentityInner {
             key: Key::from_master(key),
+            key_v2: Key::from_master(&key_v2),
```

Comment:
why do we need second key/

Reply:
The old value was essentially unstructured so I did not want to depend on 'does it parse as JSON'. That way it is backward compatible even if username happened to be parsable as JSON (I asked on [crypto.SE](https://crypto.stackexchange.com/questions/62108/is-using-derived-keys-for-hkdf-aes-gcm-safe) before writing the code).

Reply:
Can we just enable new identity and handle backward compatibility via parse error. I’d like to drop old identity in several versions.

Reply:
> Can we just enable new identity and handle backward compatibility via parse error.

At that point this depend on how conservative we want to be with what could have been stored in identity. Before this was essentially a free-form UTF-8 string so users might've store JSON.

I'm usually prefer to err on side of caution but it can be handled by parse error as well. It would just produce very strange results when user rollbacks the actix-web to previous version.

Note that current design does not preclude removing second key - it just requires that pre-processing of key is kept.

> I’d like to drop old identity in 3-4 versions.

How would you like to handle the transition? Currently you can rolllback the update of `actix-web` and, as long as you don't use new features, the cookies keep working. Assuming we have two types of keys you can either want to switch v2 cookies immediately or:

 - Introduce v2 cookies for new functionality in version X keep legacy cookies for plain type
 - Switch to v2 cookies everywhere and keep legacy cookies for parsing only
 - Remove legacy cookies